### PR TITLE
Make key pair related functions, including test

### PR DESCRIPTION
**Pull request by Baescott**
PR for making `delete_redundant_key_pairs`. This finds key pairs currently not using in each active instances and delete these pairs.

### CODE REVIEW

File: baram/ec2_manager.py [hunk 1]
```diff
@@ -13,6 +13,50 @@ def list_security_groups(self):
...
+            print(f"{number_of_key_pairs_redundant} redundant key pairs has deleted")
+
+    def list_redundant_key_pairs(self):
+        """
+        Describes all disused key pairs
+
+        :return: KeyName
+        """
+        key_pairs_total = self.list_key_pairs()
+        key_pairs_using = [instance['KeyName'] for instance in self.list_instances()]
```
Comment (lks21c) [inline baram/ec2_manager.py]: only running instances should be retrieved using instance state filter.

File: baram/ec2_manager.py [hunk 2]
```diff
@@ -13,6 +13,50 @@ def list_security_groups(self):
...
+
+    def list_key_pairs(self):
+        """
+        Describes all key pairs
+
+        :return: KeyName
+        """
+        key_pairs = self.cli.describe_key_pairs()['KeyPairs']
+        result = [key_pair['KeyName'] for key_pair in key_pairs]
+        return set(result)
```
Comment (lks21c) [inline baram/ec2_manager.py]: I prefer one single line

File: baram/ec2_manager.py [hunk 3]
```diff
@@ -13,6 +13,50 @@ def list_security_groups(self):
...
+        Delete redundant key pairs (i.e. not related to any instances)
+        """
+        key_pairs_redundant = self.list_redundant_key_pairs()
+        number_of_key_pairs_redundant = len(key_pairs_redundant)
+        if number_of_key_pairs_redundant == 0:
+            print("There's no redundant key pair")
+        else:
+            for key_pair in key_pairs_redundant:
+                self.cli.delete_key_pair(KeyName=key_pair)
+            print(f"{number_of_key_pairs_redundant} redundant key pairs has deleted")
```
Comment (lks21c) [inline baram/ec2_manager.py]: please don't use print in production code. Instead, use log manager.

**Review by lks21c — CHANGES_REQUESTED** [review]

File: baram/ec2_manager.py (hunk 3 above)
Comment (Baescott) [reply baram/ec2_manager.py]: Yes. It'll be reflected to [this](https://github.com/lks21c/baram/pull/31) shortly.

**Comment by Baescott** [comment]
@lks21c  Your comments has been reflected.

File: baram/ec2_manager.py [hunk 4]
```diff
@@ -1,90 +1,141 @@
...
         return self.cli.describe_security_groups()['SecurityGroups']
 
+    def list_instances(self):
+        """
+        Describes all existing instances.
+
+        :return:
+        """
+        instances = self.cli.describe_instances()['Reservations']
+        return [instance['Instances'][0] for instance in instances]
```
Comment (lks21c) [inline baram/ec2_manager.py]: What if there's no instance? I think it's possible depending on each AWS environment.

**Review by lks21c — CHANGES_REQUESTED** [review]

File: baram/ec2_manager.py (hunk 4 above)
Comment (Baescott) [reply baram/ec2_manager.py]: Should be more defensive. I'll apply `try & except` .

**Comment by Baescott** [comment]
@lks21c Your comments has been reflected.

File: baram/ec2_manager.py [hunk 5]
```diff
@@ -1,105 +1,154 @@
...
-        '''
+        """
         Describes the specified security groups or all of your security groups.
 
         :return: SecurityGroups
-        '''
+        """
         return self.cli.describe_security_groups()['SecurityGroups']
 
+    def list_specific_status_instances(self, status: str = 'running'):
```
Comment (lks21c) [inline baram/ec2_manager.py]: how about `list_instances_with_status`?

File: baram/ec2_manager.py [hunk 6]
```diff
@@ -1,105 +1,154 @@
...
+
+    def delete_redundant_key_pairs(self):
+        """
+        Delete redundant key pairs (i.e. not related to any instances)
+        """
+        key_pairs_redundant = self.list_redundant_key_pairs()
+        try:
+            for key_pair in key_pairs_redundant:
+                self.cli.delete_key_pair(KeyName=key_pair)
+                self.logger.info('info')
```
Comment (lks21c) [inline baram/ec2_manager.py]: ?

File: baram/ec2_manager.py [hunk 7]
```diff
@@ -1,105 +1,154 @@
...
+        except:
+            traceback.print_exc()
+
+    def list_redundant_key_pairs(self):
```
Comment (lks21c) [inline baram/ec2_manager.py]: This should be `list_unused_key_pairs`

**Review by lks21c — CHANGES_REQUESTED** [review]

**Comment by lks21c** [comment]
@Baescott please let me know after you revise code.

File: baram/ec2_manager.py (hunk 5 above)
Comment (Baescott) [reply baram/ec2_manager.py]: That name would be better.

File: baram/ec2_manager.py (hunk 6 above)
Comment (Baescott) [reply baram/ec2_manager.py]: Oh, that should be 'key has deleted' 😅

File: baram/ec2_manager.py (hunk 7 above)
Comment (Baescott) [reply baram/ec2_manager.py]: That would be better name.

**Comment by Baescott** [comment]
@lks21c Your comments has been reflected

File: tests/test_ec2_manager.py [hunk 8]
```diff
@@ -16,6 +16,22 @@ def test_list_security_groups(em):
     assert em.list_security_groups()
 
 
+def test_list_instances_with_status(em):
+    pprint(em.list_instances_with_status())
+
+
+def test_delete_unused_key_pairs(em):
+    em.delete_unused_key_pairs()
```
Comment (alprinf) [inline tests/test_ec2_manager.py:192]: I think there should be a statement to check if unused key pairs are properly deleted

**Review by alprinf — CHANGES_REQUESTED** [review]

File: tests/test_ec2_manager.py (hunk 8 above)
Comment (Baescott) [reply tests/test_ec2_manager.py:192]: As you said, KMS related unit tests need more assertions and 'Given When Then' formatting. Your comment will be reflected soon

**Review by lks21c — APPROVED** [review]

**Pull request merged**